src/thermo/RedlichKwongMFTP.cpp

```
@@ -566,18 +578,85 @@ void RedlichKwongMFTP::initThermoXML(XML_Node& phaseNode, const std::string& id)
                 XML_Node& xmlACChild = acNode.child(i);
                 if (ba::iequals(xmlACChild.name(), "purefluidparameters")) {
                     readXMLPureFluid(xmlACChild);
-                } else if (ba::iequals(xmlACChild.name(), "crossfluidparameters")) {
+                    nParams += 1;
+                }
+            }
+
+            // If any species exist which have undefined pureFluidParameters, search the database
+            //     in 'critProperties.xml' to find critical temperature and pressure and calculate
+            //     a and b based on that.
+            if (m_kk > nParams) // activity coefficients of all species are not listed in CTI file. Look into the database.
+            {
+                // Get the array of species names:
+                XML_Node& sparrays = phaseNode.child("speciesArray");
+                std::vector<std::string> spnames;
+                getStringArray(sparrays, spnames);
+
+                // Loop through all species in the CTI file
+                size_t iSpecies = 0;
+
+                for (size_t i = 0; i < m_kk; i++)
+                {
+                    string iName = ba::to_lower_copy(spnames[i]);
+
+                    // Get the index of the species
+                    iSpecies = speciesIndex(iName);
+
+                    // Check if a and b are already populated (only the diagonal elements of a).
+                    size_t counter = iSpecies + m_kk * iSpecies;
+
+                    // If not, then search the database:
+                    if (a_coeff_vec(0, counter) == -1 ||
+                        b_vec_Curr_[iSpecies] == -1) {
+
+                        doublereal *coeffArray;
+
+                        // Search the database for the species name and calculate coefficients
+                        // a and b, from critical properties:
+                        // coeffArray[0] = a0, coeffArray[1] = b;
+                        coeffArray = getCoeff(iName);
```

If not finding coefficients for a species either in the phase definition or the database is considered an error, would it make more sense to just have `getCoeff` raise an exception if the species is missing, rather than returning a special value and then checking for that value here? I think this would also simplify the implementation of `getCoeff`.